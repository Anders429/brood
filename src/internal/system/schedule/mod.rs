pub(crate) mod raw_task;
pub(crate) mod task;
