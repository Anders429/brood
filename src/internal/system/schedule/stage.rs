use crate::{
    internal::system::schedule::sendable::SendableWorld,
    registry::Registry,
    system::{
        schedule::stage::{Null, Stage},
        ParSystem, System,
    },
};

pub trait StagesSeal<'a>: Send {
    fn run<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn defer<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn run_current<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn run_continuing<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;

    fn flush<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a;
}

impl<'a> StagesSeal<'a> for Null {
    fn run<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn defer<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn run_current<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn run_continuing<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }

    fn flush<R>(&mut self, _world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
    }
}

impl<'a, S, P, L> StagesSeal<'a> for (Stage<S, P>, L)
where
    S: System<'a> + Send,
    P: ParSystem<'a> + Send,
    L: StagesSeal<'a>,
{
    fn run<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        self.defer(world);
        self.run_current(world);
    }

    fn defer<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match self.0 {
            Stage::Start(_) | Stage::Flush => {
                self.1.run(world);
            }
            Stage::Continue(_) => {
                self.1.defer(world);
            }
        }
    }

    fn run_current<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match &mut self.0 {
            Stage::Start(task) => {
                task.run(world);
            }
            Stage::Continue(task) => {
                rayon::join(
                    || {
                        self.1.run_continuing(world);
                    },
                    || {
                        task.run(world);
                    },
                );
            }
            Stage::Flush => {
                self.1.flush(world);
            }
        }
    }

    fn run_continuing<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match &mut self.0 {
            Stage::Start(task) => {
                task.run(world);
            }
            Stage::Continue(task) => {
                rayon::join(
                    || {
                        self.1.run_continuing(world);
                    },
                    || {
                        task.run(world);
                    },
                );
            }
            Stage::Flush => {}
        }
    }

    fn flush<R>(&mut self, world: SendableWorld<'a, R>)
    where
        R: Registry + 'a,
    {
        match &mut self.0 {
            Stage::Start(task) => {
                task.flush(world);
            }
            Stage::Continue(task) => {
                self.1.flush(world);
                task.flush(world);
            }
            Stage::Flush => {}
        }
    }
}
