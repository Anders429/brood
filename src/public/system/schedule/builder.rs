use crate::{
    component::Component,
    entity,
    internal::{
        query::claim::Claim,
        system::schedule::task::Task,
    },
    query::{
        filter, result, view,
        view::{View, Views},
    },
    registry::Registry,
    system::{
        schedule::{
            stage,
            stage::{Stage, Stages},
            Schedule,
        },
        ParSystem, System,
    },
    world::World,
};
use core::{any::TypeId, hint::unreachable_unchecked};
use hashbrown::HashSet;

pub enum RawTask<S, P> {
    Task(Task<S, P>),
    Flush,
}

pub struct Null;

pub trait RawTasks<'a> {
    type Stages: Stages<'a>;

    fn into_stages(
        self,
        mutable_claims: &mut HashSet<TypeId>,
        immutable_claims: &mut HashSet<TypeId>,
        mutable_buffer: &mut HashSet<TypeId>,
        immutable_buffer: &mut HashSet<TypeId>,
    ) -> Self::Stages;
}

impl<'a> RawTasks<'a> for Null {
    type Stages = stage::Null;

    fn into_stages(
        self,
        _mutable_claims: &mut HashSet<TypeId>,
        _immutable_claims: &mut HashSet<TypeId>,
        _mutable_buffer: &mut HashSet<TypeId>,
        _immutable_buffer: &mut HashSet<TypeId>,
    ) -> Self::Stages {
        stage::Null
    }
}

impl<'a, S, P, T> RawTasks<'a> for (RawTask<S, P>, T)
where
    S: System<'a> + Send,
    P: ParSystem<'a> + Send,
    T: RawTasks<'a>,
{
    type Stages = (Stage<S, P>, T::Stages);

    fn into_stages(
        self,
        mutable_claims: &mut HashSet<TypeId>,
        immutable_claims: &mut HashSet<TypeId>,
        mutable_buffer: &mut HashSet<TypeId>,
        immutable_buffer: &mut HashSet<TypeId>,
    ) -> Self::Stages {
        let prev_stages = self.1.into_stages(
            mutable_claims,
            immutable_claims,
            mutable_buffer,
            immutable_buffer,
        );

        match self.0 {
            RawTask::Task(task) => {
                mutable_buffer.clear();
                immutable_buffer.clear();

                // Identify this stage's claims on components.
                S::Views::claim(mutable_buffer, immutable_buffer);
                P::Views::claim(mutable_buffer, immutable_buffer);

                // Helper function to check whether the intersection betwen two sets is nonempty.
                fn intersects(a: &HashSet<TypeId>, b: &HashSet<TypeId>) -> bool {
                    a.intersection(b).next().is_some()
                }

                // If the claims are incompatible, a new stage must begin.
                //
                // Claims are incompatible if an immutable claim is made on a component already
                // mutable claimed, or if a mutable claim is made on a component already claimed at
                // all.
                if intersects(immutable_buffer, mutable_claims)
                    || intersects(mutable_buffer, mutable_claims)
                    || intersects(mutable_buffer, immutable_claims)
                {
                    (Stage::Start(task), prev_stages)
                } else {
                    (Stage::Continue(task), prev_stages)
                }
            }
            RawTask::Flush => {
                mutable_claims.clear();
                immutable_claims.clear();
                (Stage::Flush, prev_stages)
            }
        }
    }
}

impl<'a> System<'a> for Null {
    type Filter = filter::None;
    type Views = view::Null;

    fn run<R>(&mut self, _query_results: result::Iter<'a, R, Self::Filter, Self::Views>)
    where
        R: Registry + 'a,
    {
        unsafe { unreachable_unchecked() }
    }

    fn world_post_processing<R>(&mut self, _world: &mut World<R>)
    where
        R: Registry,
    {
        unsafe { unreachable_unchecked() }
    }
}

impl<'a> ParSystem<'a> for Null {
    type Filter = filter::None;
    type Views = view::Null;

    fn run<R>(&mut self, _query_results: result::ParIter<'a, R, Self::Filter, Self::Views>)
    where
        R: Registry + 'a,
    {
        unsafe { unreachable_unchecked() }
    }

    fn world_post_processing<R>(&mut self, _world: &mut World<R>)
    where
        R: Registry,
    {
        unsafe { unreachable_unchecked() }
    }
}

pub struct Builder<T> {
    raw_tasks: T,
}

impl Builder<Null> {
    pub(super) fn new() -> Self {
        Self { raw_tasks: Null }
    }
}

impl<'a, T> Builder<T>
where
    T: RawTasks<'a>,
{
    pub fn system<S>(self, system: S) -> Builder<(RawTask<S, Null>, T)>
    where
        S: System<'a>,
    {
        Builder::<(RawTask<S, Null>, T)> {
            raw_tasks: (RawTask::Task(Task::Seq(system)), self.raw_tasks),
        }
    }

    pub fn par_system<S>(self, par_system: S) -> Builder<(RawTask<Null, S>, T)>
    where
        S: ParSystem<'a>,
    {
        Builder::<(RawTask<Null, S>, T)> {
            raw_tasks: (RawTask::Task(Task::Par(par_system)), self.raw_tasks),
        }
    }

    pub fn flush(self) -> Builder<(RawTask<Null, Null>, T)> {
        Builder::<(RawTask<Null, Null>, T)> {
            raw_tasks: (RawTask::Flush, self.raw_tasks),
        }
    }

    pub fn build(self) -> Schedule<T::Stages> {
        Schedule {
            stages: self.raw_tasks.into_stages(
                &mut HashSet::new(),
                &mut HashSet::new(),
                &mut HashSet::new(),
                &mut HashSet::new(),
            ),
        }
    }
}
