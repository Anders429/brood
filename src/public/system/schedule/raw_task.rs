use crate::{
    internal::system::schedule::raw_task::RawTasksSeal,
    system::{schedule::task::Task, ParSystem, System},
};

pub enum RawTask<S, P> {
    Task(Task<S, P>),
    Flush,
}

pub struct Null;

pub trait RawTasks<'a>: RawTasksSeal<'a> {}

impl<'a> RawTasks<'a> for Null {}

impl<'a, S, P, T> RawTasks<'a> for (RawTask<S, P>, T)
where
    S: System<'a> + Send,
    P: ParSystem<'a> + Send,
    T: RawTasks<'a>,
{
}
