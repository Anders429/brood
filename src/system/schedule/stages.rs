use crate::{
    archetype,
    hlist::define_null,
    registry::Registry,
    system::schedule::{
        sendable::SendableWorld,
        Stage,
    },
    World,
};
use fnv::FnvBuildHasher;
use hashbrown::HashMap;

define_null!();

/// The stages within a schedule.
pub trait Stages<'a, R, FI, VI, P, I, Q>: Send
where
    R: Registry,
{
    /// A list of booleans indicating whether each task within the first stage has already been run.
    type HasRun: Send;

    /// Run all of the stages, parallelizing as much work as possible.
    ///
    /// The parallelization strategy involves two parts:
    ///
    /// 1. Compile-time scheduling: at compile time, tasks are split into stages, where all tasks
    /// in a stage can always be run in parallel with each other, no matter the `World`.
    /// 2. Run-time optimization: at run-time, component claims on archetype tables within the
    /// `World` are tracked when scheduling a single stage. Then, any tasks within the next stage
    /// whose borrowed components do not interfere with the tasks in the current stage's dynamic
    /// claims are run as well.
    fn run(&mut self, world: &mut World<R>, has_run: Self::HasRun);

    /// Attempt to run as many tasks within the first stage in the list as possible as add-ons to
    /// the previous stage.
    ///
    /// `borrowed_archetypes` contains a set of dynamic claims that are already borrowed by the
    /// previous stage. This method respects those claims when evaluating whether new tasks can be
    /// executed.
    ///
    /// # Safety
    /// `borrowed_archetypes` must accurately represent the dynamic claims already made on the
    /// component columns within `world`.
    unsafe fn run_add_ons(
        &mut self,
        world: SendableWorld<R>,
        borrowed_archetypes: HashMap<archetype::IdentifierRef<R>, R::Claims, FnvBuildHasher>,
    ) -> Self::HasRun;

    /// Creates a new default set of booleans to indicate that each task within the first stage has
    /// not been run.
    fn new_has_run() -> Self::HasRun;
}

impl<R> Stages<'_, R, Null, Null, Null, Null, Null> for Null
where
    R: Registry,
{
    type HasRun = Null;

    fn run(&mut self, _world: &mut World<R>, _has_run: Self::HasRun) {}

    unsafe fn run_add_ons(
        &mut self,
        _world: SendableWorld<R>,
        _borrowed_archetypes: HashMap<archetype::IdentifierRef<R>, R::Claims, FnvBuildHasher>,
    ) -> Self::HasRun {
        Null
    }

    fn new_has_run() -> Self::HasRun {
        Null
    }
}

impl<'a, R, T, U, FI, FIS, VI, VIS, P, PS, I, IS, Q, QS>
    Stages<'a, R, (FI, FIS), (VI, VIS), (P, PS), (I, IS), (Q, QS)> for (T, U)
where
    R: Registry,
    T: Stage<'a, R, FI, VI, P, I, Q>,
    U: Stages<'a, R, FIS, VIS, PS, IS, QS>,
{
    type HasRun = T::HasRun;

    fn run(&mut self, world: &mut World<R>, has_run: Self::HasRun) {
        // Each stage is run sequentially. The tasks within a stage are parallelized.
        let next_has_run = self.0.run(
            // SAFETY: The pointer provided here is unique, being created from a mutable reference.
            unsafe { SendableWorld::new(world) },
            HashMap::default(),
            has_run,
            &mut self.1,
        );
        self.1.run(world, next_has_run);
    }

    unsafe fn run_add_ons(
        &mut self,
        world: SendableWorld<R>,
        borrowed_archetypes: HashMap<archetype::IdentifierRef<R>, R::Claims, FnvBuildHasher>,
    ) -> Self::HasRun {
        // SAFETY: The safety contract of this method call is upheld by the safety contract of this
        // method.
        unsafe { self.0.run_add_ons(world, borrowed_archetypes) }
    }

    fn new_has_run() -> Self::HasRun {
        T::new_has_run()
    }
}
