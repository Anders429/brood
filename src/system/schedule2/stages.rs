use crate::{
    World,
    hlist::define_null,
    registry::Registry,
    system::schedule2::{Stage, sendable::SendableWorld},
};

define_null!();

pub trait Stages<R, FI, VI, P, I, Q> where R: Registry {
    fn run(&mut self, world: &mut World<R>);
}

impl<R> Stages<R, Null, Null, Null, Null, Null> for Null where R: Registry {
    fn run(&mut self, _world: &mut World<R>) {}
}

impl<R, T, U, FI, FIS, VI, VIS, P, PS, I, IS, Q, QS> Stages<R, (FI, FIS), (VI, VIS), (P, PS), (I, IS), (Q, QS)> for (T, U)
where
    R: Registry,
    T: Stage<R, FI, VI, P, I, Q>,
    U: Stages<R, FIS, VIS, PS, IS, QS>,
{
    fn run(&mut self, world: &mut World<R>) {
        // Each stage is run sequentially. The tasks within a stage are parallelized.
        self.0.run(
            // SAFETY: The pointer provided here is unique, being created from a mutable reference.
            unsafe { SendableWorld::new(world) },
        );
        self.1.run(world);
    }
}
