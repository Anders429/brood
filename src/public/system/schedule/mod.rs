pub mod stage;
pub mod task;

mod builder;

pub use builder::Builder;

use crate::{
    internal::system::schedule::{sendable::SendableWorld, raw_task}, registry::Registry,
    world::World,
};
use stage::Stages;

pub struct Schedule<S> {
    stages: S,
}

impl Schedule<stage::Null> {
    pub fn builder() -> Builder<raw_task::Null> {
        Builder::new()
    }
}

impl<'a, S> Schedule<S>
where
    S: Stages<'a>,
{
    pub(crate) fn run<R>(&mut self, world: &'a mut World<R>)
    where
        R: Registry,
    {
        self.stages.run(SendableWorld(world));
    }
}
