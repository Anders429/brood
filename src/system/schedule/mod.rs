//! A list of [`System`]s to be run in stages.
//!
//! [`Schedule`]s are created using a builder pattern. `System`s are provided in the desired order
//! they are to be run, and the stages in which those `System`s are run is automatically derived.
//!
//! The advantage of defining a `Schedule` is that `System`s are allowed to be run in parallel as
//! long as their [`Views`] can be borrowed simultaneously.
//!
//! # Example
//! The below example will execute both `SystemA` and `SystemB` in parallel, since their views can
//! be borrowed simultaneously.
//!
//! ``` rust
//! use brood::{
//!     query::{
//!         filter,
//!         filter::Filter,
//!         result,
//!         views,
//!     },
//!     registry::ContainsQuery,
//!     system::{
//!         Schedule,
//!         System,
//!     },
//! };
//!
//! // Define components.
//! struct Foo(usize);
//! struct Bar(bool);
//! struct Baz(f64);
//!
//! struct SystemA;
//!
//! impl<'a> System<'a> for SystemA {
//!     type Views = views!(&'a mut Foo, &'a Bar);
//!     type Filter = filter::None;
//!
//!     fn run<R, FI, VI, P, I, Q>(
//!         &mut self,
//!         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
//!     ) where
//!         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
//!     {
//!         for result!(foo, bar) in query_results {
//!             // Do something...
//!         }
//!     }
//! }
//!
//! struct SystemB;
//!
//! impl<'a> System<'a> for SystemB {
//!     type Views = views!(&'a mut Baz, &'a Bar);
//!     type Filter = filter::None;
//!
//!     fn run<R, FI, VI, P, I, Q>(
//!         &mut self,
//!         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
//!     ) where
//!         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
//!     {
//!         for result!(baz, bar) in query_results {
//!             // Do something...
//!         }
//!     }
//! }
//!
//! let schedule = Schedule::builder().system(SystemA).system(SystemB).build();
//! ```
//!
//! [`Schedule`]: crate::system::schedule::Schedule
//! [`System`]: crate::system::System
//! [`Views`]: crate::query::view::Views

pub mod raw_task;
pub mod stage;

pub(crate) mod task;

mod sendable;

mod builder;

pub use builder::Builder;
pub use stage::stages;

use crate::{
    registry::Registry,
    world::World,
};
use sendable::SendableWorld;
use stage::Stages;

/// A list of [`System`]s to be run in stages.
///
/// The `System`s that make up a `Schedule` are organized into [`Stages`] on creation. `System`s
/// that can be run in parallel are done so. See the documentation for the [`schedule::Builder`]
/// for more information about creating a `Schedule`.
///
/// # Example
/// The below example will execute both `SystemA` and `SystemB` in parallel, since their views can
/// be borrowed simultaneously.
///
/// ``` rust
/// use brood::{
///     query::{
///         filter,
///         filter::Filter,
///         result,
///         views,
///     },
///     registry::ContainsQuery,
///     system::{
///         Schedule,
///         System,
///     },
/// };
///
/// // Define components.
/// struct Foo(usize);
/// struct Bar(bool);
/// struct Baz(f64);
///
/// struct SystemA;
///
/// impl<'a> System<'a> for SystemA {
///     type Views = views!(&'a mut Foo, &'a Bar);
///     type Filter = filter::None;
///
///     fn run<R, FI, VI, P, I, Q>(
///         &mut self,
///         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
///     ) where
///         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
///     {
///         for result!(foo, bar) in query_results {
///             // Do something...
///         }
///     }
/// }
///
/// struct SystemB;
///
/// impl<'a> System<'a> for SystemB {
///     type Views = views!(&'a mut Baz, &'a Bar);
///     type Filter = filter::None;
///
///     fn run<R, FI, VI, P, I, Q>(
///         &mut self,
///         query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
///     ) where
///         R: ContainsQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
///     {
///         for result!(baz, bar) in query_results {
///             // Do something...
///         }
///     }
/// }
///
/// let schedule = Schedule::builder().system(SystemA).system(SystemB).build();
/// ```
///
/// [`schedule::Builder`]: crate::system::schedule::Builder
/// [`Stages`]: crate::system::schedule::stage::Stages
/// [`System`]: crate::system::System
#[cfg_attr(doc_cfg, doc(cfg(feature = "rayon")))]
pub struct Schedule<S> {
    stages: S,
}

impl Schedule<stage::Null> {
    /// Creates a [`schedule::Builder`] to construct a new `Schedule`.
    ///
    /// # Example
    /// ``` rust
    /// use brood::system::Schedule;
    ///
    /// let builder = Schedule::builder();
    /// // Add systems to the builder.
    /// let schedule = builder.build();
    /// ```
    ///
    /// [`schedule::Builder`]: crate::system::schedule::Builder
    #[must_use]
    pub fn builder() -> Builder<raw_task::Null> {
        Builder::new()
    }
}

impl<'a, S> Schedule<S> {
    pub(crate) fn run<R, SFI, SVI, PFI, PVI, SP, SI, SQ, PP, PI, PQ>(
        &mut self,
        world: &'a mut World<R>,
    ) where
        R: Registry,
        S: Stages<'a, R, SFI, SVI, PFI, PVI, SP, SI, SQ, PP, PI, PQ>,
    {
        self.stages.run(
            // SAFETY: The pointer provided here is unique, being created from a mutable reference.
            unsafe { SendableWorld::new(world) },
        );
    }
}
