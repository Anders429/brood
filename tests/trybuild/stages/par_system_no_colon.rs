use brood::system::schedule::stages;

type Stages = stages!{
    par_system,
};

fn main() {}
