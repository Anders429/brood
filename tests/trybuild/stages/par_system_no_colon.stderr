error: no rules expected the token `,`
 --> tests/trybuild/stages/par_system_no_colon.rs:4:15
  |
4 |     par_system,
  |               ^ no rules expected this token in macro call
